File: Contest/Graph/23.cpp
#include <iostream>
#include <vector>

using namespace std;

int n, m;
vector<int> adj[1001];    // Adjacency list for the original graph
vector<int> t_adj[1001];  // Adjacency list for the transposed graph
int visited[1001] = {};     // Visited array
vector<int> v, v2;          // Vectors to store traversal order

// DFS for the original graph
void DFS(int i) {
    visited[i] = 1;  // Mark node as visited
    v.push_back(i);  // Store the node in traversal order
    for (int j : adj[i]) {
        if (visited[j] == 0) {
            DFS(j);  // Continue DFS if not visited
        }
    }
}
// DFS for the transposed graph
void DFS2(int i) {
    visited[i] = 1;  // Mark node as visited
    v2.push_back(i);  // Store the node in traversal order
    for (int j : t_adj[i]) {
        if (visited[j] == 0) {
            DFS2(j);  // Continue DFS on the transposed graph if not visited
        }
    }
}

int main() {
    cin >> n >> m;
    
    // Read the edges and build both the original and transposed graphs
    for (int i = 0; i < m; i++) {
        int x, y;
        cin >> x >> y;
        adj[x].push_back(y);  // Original graph: edge from x to y
        t_adj[y].push_back(x);  // Transposed graph: edge from y to x
    }

    // Run DFS on the original graph from node 1
    DFS(1);

    // Check if there are any unvisited nodes
    for (int i = 2; i <= n; i++) {
        if (visited[i] == 0) {
            cout << "0\n";  // If any node is unvisited, the graph is not strongly connected
            return 0;
        }
    }

    // Reset visited array for the transposed graph DFS
    fill(visited, visited + 1001, 0);

    // Run DFS on the transposed graph from node 1
    DFS2(1);

    // Check again if there are any unvisited nodes in the transposed graph
    for (int i = 2; i <= n; i++) {
        if (visited[i] == 0) {
            cout << "0\n";  // If any node is unvisited in the transposed graph, not strongly connected
            return 0;
        }
    }

    // If all nodes are visited in both the original and transposed graph DFS, the graph is strongly connected
    cout << "1\n";
    return 0;
}

File: SinhTest/Graph/23.py
import random
import os

MAX_VERTICES = 1000
MAX_EDGES = 1000

# DFS for the original graph
def dfs(graph, visited, node):
    visited[node] = 1
    for neighbor in graph[node]:
        if visited[neighbor] == 0:
            dfs(graph, visited, neighbor)

# DFS for the transposed graph
def dfs_transposed(graph, visited, node):
    visited[node] = 1
    for neighbor in graph[node]:
        if visited[neighbor] == 0:
            dfs_transposed(graph, visited, neighbor)

# Check if the graph is strongly connected using Kosaraju's algorithm
def is_strongly_connected(n, graph, transposed_graph):
    visited = [0] * (n + 1)
    
    # Run DFS on the original graph starting from node 1
    dfs(graph, visited, 1)

    # Check if all nodes were visited in the original graph
    if any(visited[i] == 0 for i in range(2, n + 1)):
        return False

    # Reset visited array and run DFS on the transposed graph
    visited = [0] * (n + 1)
    dfs_transposed(transposed_graph, visited, 1)

    # Check if all nodes were visited in the transposed graph
    if any(visited[i] == 0 for i in range(2, n + 1)):
        return False

    return True

# Generate a random test case and check if the graph is strongly connected
def generate_test_case(test_num):
    # Random number of vertices and edges
    n = random.randint(2, MAX_VERTICES)
    m = random.randint(1, min(n * (n - 1), MAX_EDGES))

    # Generate edges
    edges = set()
    while len(edges) < m:
        u = random.randint(1, n)
        v = random.randint(1, n)
        if u != v:  # Ensure no self-loops
            edges.add((u, v))

    graph = [[] for _ in range(n + 1)]
    transposed_graph = [[] for _ in range(n + 1)]
    
    # Build the adjacency list for the original and transposed graphs
    for u, v in edges:
        graph[u].append(v)
        transposed_graph[v].append(u)

    # Write input file
    input_filename = f"input/input{test_num:03d}.txt"
    os.makedirs('input', exist_ok=True)
    with open(input_filename, 'w') as input_file:
        input_file.write(f"{n} {m}\n")
        for u, v in edges:
            input_file.write(f"{u} {v}\n")
    
    # Check if the graph is strongly connected and write output
    output_filename = f"output/output{test_num:03d}.txt"
    os.makedirs('output', exist_ok=True)
    with open(output_filename, 'w') as output_file:
        if is_strongly_connected(n, graph, transposed_graph):
            output_file.write("1\n")
        else:
            output_file.write("0\n")

# Generate 50 test cases and update the terminal
for i in range(50):  # Generate test cases from 0 to 49
    generate_test_case(i)
    os.system('cls' if os.name == 'nt' else 'clear')  # Clear the screen
    print(f"{i + 1}/50 test cases generated.")

print("50 test cases have been written to input/input000.txt - input/input049.txt and output/output000.txt - output/output049.txt.")
